yjit/src/codegen.rs

```diff
@@ -1326,6 +1326,48 @@ fn guard_object_is_array(
     asm.jne(side_exit.as_side_exit());
 }
 
+/// This guards that a special flag is not set on a hash.
+/// By passing a hash with this flag set as the last argument
+/// in a splat call, you can change the way keywords are handled
+/// to behave like ruby 2. We don't currently support this.
+fn guard_object_is_not_ruby2_keyword_hash(
+    asm: &mut Assembler,
+    object_opnd: Opnd,
+    side_exit: CodePtr,
+) {
+    asm.comment("guard object is not ruby2 keyword hash");
+
+    let label = asm.new_label("exit");
+    asm.test(object_opnd, (RUBY_IMMEDIATE_MASK as u64).into());
+    asm.jnz(label);
+
+    asm.cmp(object_opnd, Qfalse.into());
+    asm.je(label);
+
+    let flags_opnd = Opnd::mem(
+        8 * SIZEOF_VALUE as u8,
+        object_opnd,
+        RUBY_OFFSET_RBASIC_FLAGS,
+    );
+    let type_opnd = asm.and(flags_opnd, (RUBY_T_MASK as u64).into());
+
+    asm.cmp(type_opnd, (RUBY_T_HASH as u64).into());
+    asm.jne(label);
+
+    let flags_opnd = Opnd::mem(
+        8 * SIZEOF_VALUE as u8,
+        object_opnd,
+        RUBY_OFFSET_RBASIC_FLAGS,
+    );
+
+    let keyword_flag = asm.and(flags_opnd, (RHASH_PASS_AS_KEYWORDS as u64).into());
+
+    asm.cmp(keyword_flag, 0.into());
```

Comment:
Is it possible to use a `test` instruction instead of `and` + `cmp`?

Reply:
Done